source/rust_verify/src/rust_to_vir_base.rs

```diff
@@ -192,23 +192,93 @@ pub(crate) fn qpath_to_ident<'tcx>(
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct ClauseFrom<'tcx> {
+    clause: Clause<'tcx>,
+    is_self_trait_bound: bool,
+    source_id: DefId,
+    span: Span,
+}
+
+fn instantiate_pred_clauses<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    mut def_id: DefId,
+    args: rustc_middle::ty::GenericArgsRef<'tcx>,
+) -> Vec<(Option<ClauseFrom<'tcx>>, Clause<'tcx>)> {
+    // We could get the information directly like this:
+    let direct_clauses = tcx.predicates_of(def_id).instantiate(tcx, args).predicates;
```

Comment:
Why was this sufficient before, and now we need the additional information. Is it just to distinguish the `self` trait bound?

Reply:
> Why was this sufficient before, and now we need the additional information. Is it just to distinguish the `self` trait bound?

Yes, that's right.  The only place that uses the additional information (`ClauseFrom.is_self_trait_bound`) is this:

```
                    match (&mut remove_self_trait_bound, inst_bound) {
                        (Some((expected_id, self_trait_impl_path)), Some(b))
                            if *expected_id == b.source_id && b.is_self_trait_bound =>
                        {
                            assert!(self_trait_impl_path.is_none());
                            **self_trait_impl_path = Some(impl_path);
                        }
                        _ => {
                            impl_paths.push(impl_path);
                        }
                    }
```

which only matters when `remove_self_trait_bound` is Some, which is only used when handling direct calls to default method implementations.